presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultiJoinNode.java

```diff
@@ -0,0 +1,141 @@
+        this.filter = requireNonNull(filter, "filter is null");
+        this.outputSymbols = ImmutableList.copyOf(requireNonNull(outputSymbols, "outputSymbols is null"));
+
+        List<Symbol> inputSymbols = sources.stream().flatMap(source -> source.getOutputSymbols().stream()).collect(toImmutableList());
+        checkArgument(inputSymbols.containsAll(outputSymbols), "inputs do not contain all output symbols");
+    }
+
+    public Expression getFilter()
+    {
+        return filter;
+    }
+
+    public List<PlanNode> getSources()
+    {
+        return sources;
+    }
+
+    public List<Symbol> getOutputSymbols()
+    {
+        return outputSymbols;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(
+                ImmutableSet.copyOf(this.getOutputSymbols()),
+                ImmutableSet.copyOf(extractConjuncts(this.getFilter())),
+                this.getSources().stream()
+                        .map(PlanNode::getId)
+                        .collect(toImmutableSet()));
+    }
+
+    @Override
+    public boolean equals(Object other)
+    {
+        if (!(other instanceof MultiJoinNode)) {
+            return false;
+        }
+
+        MultiJoinNode otherNode = (MultiJoinNode) other;
+        return ImmutableSet.copyOf(this.getOutputSymbols()).equals(ImmutableSet.copyOf(otherNode.getOutputSymbols()))
+                && ImmutableSet.copyOf(extractConjuncts(this.getFilter())).equals(ImmutableSet.copyOf(extractConjuncts(otherNode.getFilter())))
+                && this.getSources().stream().map(PlanNode::getId).collect(toImmutableSet()).equals(otherNode.getSources().stream().map(PlanNode::getId).collect(toImmutableSet()));
+    }
+
+    static MultiJoinNode toMultiJoinNode(JoinNode joinNode, Lookup lookup)
+    {
+        return new MultiJoinNodeBuilder(joinNode, lookup).toMultiJoinNode();
+    }
+
+    private static class MultiJoinNodeBuilder
+    {
+        private final List<PlanNode> sources = new ArrayList<>();
+        private final List<Expression> filters = new ArrayList<>();
+        private final List<Symbol> outputSymbols;
+        private final Lookup lookup;
+
+        MultiJoinNodeBuilder(JoinNode node, Lookup lookup)
+        {
+            requireNonNull(node, "node is null");
+            checkState(node.getType() == INNER, "join type must be INNER");
+            this.outputSymbols = node.getOutputSymbols();
+            this.lookup = requireNonNull(lookup, "lookup is null");
+            flattenNode(node);
+        }
+
+        private void flattenNode(PlanNode node)
+        {
+            PlanNode resolved = lookup.resolve(node);
+            if (resolved instanceof JoinNode && ((JoinNode) resolved).getType() == INNER && isDeterministic(((JoinNode) resolved).getFilter().orElse(TRUE_LITERAL))) {
+                JoinNode joinNode = (JoinNode) resolved;
+                flattenNode(joinNode.getLeft());
```

Comment:
> the EO and I weren't introduced to each other

you were, it's just the name changed. Martin said he sees that as a completely new optimizer when compared to the Iterative one, with a different memo. They will most likely coexist for some time. We also agreed on a new name so that it's clear about which 'new optimizer' we're talking.

The rules you mention will have to be implemented in a different way, possibly influencing the exploratory optimizer design. Otherwise, either:
1. the exploratory optimizer (and other possible variants) won't be able to control the traversal and prune it because of rules overtaking the steer
2. we'd have to invest in recursive patterns and custom interpretation strategies of those patterns for exploratory optimizer (and other possible optimizers) waaaay beyond even I find reasonable

This, I think, is far ahead us though - we should start with simple rules and shape the API, having the more complex use cases in mind and designing the API carefully.